libraries/botframework-connector/src/auth/botFrameworkAuthentication.ts

```diff
@@ -0,0 +1,104 @@
+ */
+export abstract class BotFrameworkAuthentication {
+    /**
+     * Validate Bot Framework Protocol requests.
+     *
+     * @param activity The inbound Activity.
+     * @param authHeader The HTTP auth header.
+     */
+    abstract authenticateRequest(activity: Activity, authHeader: string): Promise<AuthenticateRequestResult>;
+
+    /**
+     * Validate Bot Framework Protocol requests.
+     *
+     * @param authHeader The HTTP auth header.
+     * @param channelIdHeader The channel ID HTTP header.
+     */
+    abstract authenticateStreamingRequest(
+        authHeader: string,
+        channelIdHeader: string
+    ): Promise<AuthenticateRequestResult>;
+
+    /**
+     * Creates a ConnectorFactory that can be used to create ConnectorClients that can use credentials from this particular Cloud Environment.
+     *
+     * @param claimsIdentity The inbound Activity's ClaimsIdentity.
+     */
+    abstract createConnectorFactory(claimsIdentity: ClaimsIdentity): ConnectorClient;
+
+    /**
+     * Creates the appropriate UserTokenClient instance.
+     *
+     * @param claimsIdentity The inbound Activity's ClaimsIdentity.
+     */
+    abstract createUserTokenClient(claimsIdentity: ClaimsIdentity): Promise<UserTokenClient>;
+
+    /**
+     * Creates a BotFrameworkClient for calling Skills.
+     */
+    createBotFrameworkClient(): BotFrameworkClient {
+        return notImplementedError();
+    }
+
+    /**
+     * Gets the originating audience from Bot OAuth scope.
+     */
+    getOriginatingAudience(): string {
+        return notImplementedError();
+    }
+
+    // TODO: Update docstring - this is a direct port from .NET
+    /**
+     * Authenticate Bot Framework Protocol request to Skills.
+     *
+     * @param authHeader The HTTP auth header in the skill request.
+     */
+    authenticateChannelRequest(authHeader: string): Promise<ClaimsIdentity> {
+        return notImplementedError();
+    }
+
+    /**
+     * Generates the appropriate callerId to write onto the Activity, this might be null.
+     *
+     * @param credentialFactory A ServiceClientCredentialsFactory to use.
+     * @param claimsIdentity The inbound claims.
+     * @param callerId The default callerId to use if this is not a skill.
+     */
+    protected async generateCallerId(
+        credentialFactory: ServiceClientCredentialsFactory,
+        claimsIdentity: ClaimsIdentity,
+        callerId: string
+    ): Promise<string> {
+        // Is the bot accepting all incoming messages?
+        if (await credentialFactory.isAuthenticationDisabled()) {
+            // Return null so that the callerId is cleared.
+            return null;
+        }
+
+        // Is the activity from another bot?
+        return SkillValidation.isSkillClaim(claimsIdentity.claims)
+            ? `${CallerIdConstants.BotToBotPrefix}${JwtTokenValidation.getAppIdFromClaims(claimsIdentity.claims)}`
+            : callerId;
+    }
+}
+
+const notImplementedError = () => {
```

Comment:
I'm not sure if this is inferred by `tsc` or not, but when you have a method that is used purely to throw an error you can use `: never` as the return type. That might make it so you don't need to do `return notImplementedError()` above. Also - are these stack traces reasonable?

Reply:
Good catch on both, I've simplified the error code to address both issues.